annotation-processor/src/main/java/org/hibernate/validator/ap/checks/annotationparameters/GroupSequenceCheck.java

```diff
@@ -0,0 +1,150 @@
+	private Types typeUtils;
+
+	public GroupSequenceCheck(AnnotationApiHelper annotationApiHelper, Types typeUtils) {
+		super( annotationApiHelper, JAVAX_VALIDATION_GROUP_SEQUENCE );
+		this.typeUtils = typeUtils;
+	}
+
+	@Override
+	protected Set<ConstraintCheckIssue> doCheck(Element element, AnnotationMirror annotation) {
+		List<? extends AnnotationValue> annotationValue = annotationApiHelper.getAnnotationArrayValue( annotation, "value" );
+
+		TypeElement annotatedElement = (TypeElement) element;
+
+		boolean classForRedefiningGroupSequencePresent = false;
+
+		for ( AnnotationValue value : annotationValue ) {
+			// 1. The class list contains only interface
+			TypeMirror typeMirror = (TypeMirror) value.getValue();
+			boolean isClassAndSameAsAnnotatedElement = isClassForRedefiningGroupSequence( annotatedElement, typeMirror );
+
+			if ( !annotationApiHelper.isInterface( typeMirror ) && !isClassAndSameAsAnnotatedElement ) {
+				return CollectionHelper.asSet( ConstraintCheckIssue.error(
+						element, annotation, "INVALID_GROUP_SEQUENCE_VALUE_NOT_INTERFACES_ANNOTATION_PARAMETERS"
+				) );
+			}
+			// 2. the defined group sequence is expandable (no cyclic definition)
+			if ( ( isFromHierarchy( annotatedElement, typeMirror ) || isPartOfGroupSequence( annotatedElement, typeMirror ) ) && !isClassAndSameAsAnnotatedElement ) {
+				return CollectionHelper.asSet( ConstraintCheckIssue.error(
+						element, annotation, "INVALID_GROUP_SEQUENCE_VALUE_CYCLIC_DEFINITION_ANNOTATION_PARAMETERS"
+				) );
+			}
+			// 3. the class list contains the hosting bean class (for default group sequence re-definition)
+			classForRedefiningGroupSequencePresent = classForRedefiningGroupSequencePresent || isClassAndSameAsAnnotatedElement;
+		}
+
+		// if the annotated element is a class and does not contain itself in a group then it's bad group redefinition
+		if ( ElementKind.CLASS.equals( annotatedElement.getKind() ) && !classForRedefiningGroupSequencePresent ) {
+			return CollectionHelper.asSet( ConstraintCheckIssue.error(
+					element, annotation, "INVALID_GROUP_SEQUENCE_VALUE_HOSTING_BEAN_ANNOTATION_PARAMETERS"
+			) );
+		}
+
+		return Collections.emptySet();
+	}
+
+	/**
+	 * Checks if a given type element ({@link TypeElement}) is not a part of definition of a group sequence annotation declared on
+	 * a given group element {@link TypeMirror}
+	 *
+	 * @param typeElement an element to check if it is present or not in definition of a group sequence
+	 * @param groupElement an type mirror at which to look for group sequence annotation
+	 *
+	 * @return {@code true} if given element is part of group sequence definition, and {@code false} otherwise
+	 */
+	private boolean isPartOfGroupSequence(TypeElement typeElement, TypeMirror groupElement) {
+		List<? extends AnnotationValue> annotationValue = null;
+		for ( AnnotationMirror annotationMirror : typeUtils.asElement( groupElement ).getAnnotationMirrors() ) {
```

Comment:
This could be structured a bit nicer with a separate method:

    AnnotationMirror groupSequence = getGroupSequence( groupElement );
    annotationValue = annotationApiHelper.getAnnotationArrayValue( annotationMirror, "value" );